arch/arm/src/sama5/sam_classd.c

```diff
@@ -0,0 +1,2617 @@
+            break;
+        }
+    }
+
+  /* reset the classD peripheral */
+
+  classd_reset(priv);
+
+  /* Return any pending buffers in our pending queue */
+
+  nxmutex_lock(&priv->pendlock);
+  while ((apb = (struct ap_buffer_s *)dq_remfirst(&priv->pendq)) != NULL)
+    {
+      /* Release our reference to the buffer */
+
+      apb_free(apb);
+
+      /* Send the buffer back up to the previous level. */
+
+#ifdef CONFIG_AUDIO_MULTI_SESSION
+      priv->dev.upper(priv->dev.priv, AUDIO_CALLBACK_DEQUEUE, apb, OK, NULL);
+#else
+      priv->dev.upper(priv->dev.priv, AUDIO_CALLBACK_DEQUEUE, apb, OK);
+#endif
+    }
+
+  nxmutex_unlock(&priv->pendlock);
+
+  classd_returnbuffers(priv);
+
+  /* Close the message queue */
+
+  file_mq_close(&priv->mq);
+  file_mq_unlink(priv->mqname);
+
+  /* Send an AUDIO_MSG_COMPLETE message to the client */
+
+#ifdef CONFIG_AUDIO_MULTI_SESSION
+  priv->dev.upper(priv->dev.priv, AUDIO_CALLBACK_COMPLETE, NULL, OK, NULL);
+#else
+  priv->dev.upper(priv->dev.priv, AUDIO_CALLBACK_COMPLETE, NULL, OK);
+#endif
+
+  audinfo("Exit\n");
+  return NULL;
+}
+
+/****************************************************************************
+ * Name: classd_start
+ *
+ * Description:
+ *   Start the configured operation (audio streaming, volume enabled, etc.).
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_AUDIO_MULTI_SESSION
+static int classd_start(struct audio_lowerhalf_s *dev, void *session)
+#else
+static int classd_start(struct audio_lowerhalf_s *dev)
+#endif
+{
+  struct classd_dev_s *priv = (struct classd_dev_s *)dev;
+  struct sched_param sparam;
+  struct mq_attr attr;
+  pthread_attr_t tattr;
+  void *value;
+  int ret;
+
+  DEBUGASSERT(priv != NULL);
+
+  audinfo("Entry\n");
+
+  /* Create a message queue for the worker thread */
+
+  snprintf(priv->mqname, sizeof(priv->mqname), "/tmp/%" PRIXPTR,
+           (uintptr_t)priv);
+
+  attr.mq_maxmsg  = 16;
+  attr.mq_msgsize = sizeof(struct audio_msg_s);
+  attr.mq_curmsgs = 0;
+  attr.mq_flags   = 0;
+
+  ret = file_mq_open(&priv->mq, priv->mqname,
+                     O_RDWR | O_CREAT, 0644, &attr);
+  if (ret < 0)
+    {
+      /* Error creating message queue! */
+
+      auderr("ERROR: Couldn't allocate message queue\n");
+      return ret;
+    }
+
+  /* Join any old worker thread we had created to prevent a memory leak */
+
+  if (priv->threadid != 0)
+    {
+      audinfo("Joining old thread\n");
+      pthread_join(priv->threadid, &value);
+    }
+
+  /* Start our thread for sending data to the device */
+
+  pthread_attr_init(&tattr);
+  sparam.sched_priority = sched_get_priority_max(SCHED_FIFO) - 3;
+  pthread_attr_setschedparam(&tattr, &sparam);
+  pthread_attr_setstacksize(&tattr,
+                            CONFIG_AUDIO_SAMA5_CLASSD_WORKER_STACKSIZE);
+
+  audinfo("Starting worker thread\n");
+  ret = pthread_create(&priv->threadid, &tattr, classd_workerthread,
```

Comment:
I just inspected the code. Really there are many usages of pthread in drivers, especially audio. Let's keep it and if rework is needed it will be done one for all the cases